File: backend-java/src/main/java/com/bffgin/backend/auth/UserResolver.java
package com.bffgin.backend.auth;

import com.bffgin.backend.domain.TaskError;
import com.bffgin.backend.domain.User;
import com.bffgin.backend.repository.TaskRepository;

import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import java.sql.SQLException;
import java.util.Optional;

/**
 * REST/gRPCの両トランスポートが共有するuser_id解決ロジック(認証ロジックを複製しない設計)。
 * backend(Go)のresolveUserID・backend-rustのauth::resolve_user_idと同じ分岐:
 *   - ローカル(HMAC/RSA)発行のJWT: subは内部user_idそのもの → usersをidで検索
 *   - Keycloak発行のJWT: subはkeycloak_sub → user_keycloaks経由で検索
 * どちらも見つからなければUserNotProvisioned
 */
public final class UserResolver {

    private static final Logger log = LoggerFactory.getLogger(UserResolver.class);

    private final Dispatcher dispatcher;
    private final TaskRepository repository;

    public UserResolver(Dispatcher dispatcher, TaskRepository repository) {
        this.dispatcher = dispatcher;
        this.repository = repository;
    }

    public long resolve(String authorizationHeader) throws TaskError {
        if (authorizationHeader == null || !authorizationHeader.startsWith("Bearer ")) {
            throw TaskError.unauthorized();
        }
        String token = authorizationHeader.substring("Bearer ".length());

        Claims claims;
        try {
            claims = dispatcher.verify(token);
        } catch (VerifyException e) {
            throw TaskError.unauthorized();
        }

        try {
            if (Dispatcher.isLocalIssuer(claims.iss())) {
                long id;
                try {
                    id = Long.parseLong(claims.sub());
                } catch (NumberFormatException e) {
                    throw TaskError.userNotProvisioned();
                }
                Optional<User> user = repository.findUserById(id);
                if (user.isEmpty()) {
                    throw TaskError.userNotProvisioned();
                }
                log.debug("user_id resolved via local issuer: iss={} sub={} user_id={}", claims.iss(), claims.sub(), user.get().id());
                return user.get().id();
            }

            Optional<User> user = repository.findUserByKeycloakSub(claims.sub());
            if (user.isEmpty()) {
                throw TaskError.userNotProvisioned();
            }
            log.debug("user_id resolved via keycloak_sub: iss={} sub={} user_id={}", claims.iss(), claims.sub(), user.get().id());
            return user.get().id();
        } catch (SQLException e) {
            throw TaskError.internal(e.getMessage());
        }
    }
}

File: backend-kotlin/src/main/kotlin/com/bffgin/backend/auth/UserResolver.kt
package com.bffgin.backend.auth

import com.bffgin.backend.domain.TaskError
import com.bffgin.backend.repository.TaskRepository
import org.slf4j.LoggerFactory

/**
 * REST/gRPCの両トランスポートが共有するuser_id解決ロジック(認証ロジックを複製しない設計)。
 * backend(Go)のresolveUserID・backend-rust/backend-javaのUserResolverと同じ分岐:
 *   - ローカル(HMAC/RSA)発行のJWT: subは内部user_idそのもの → usersをidで検索
 *   - Keycloak発行のJWT: subはkeycloak_sub → user_keycloaks経由で検索
 * どちらも見つからなければUserNotProvisioned。
 * repositoryのメソッドは内部でwithContext(Dispatchers.IO)を使うため、ここで改めて
 * ディスパッチャを意識する必要は無い(呼び出し元は「これはsuspend fun」と知るだけでよい)
 */
class UserResolver(
    private val dispatcher: Dispatcher,
    private val repository: TaskRepository,
) {
    private val log = LoggerFactory.getLogger(UserResolver::class.java)

    suspend fun resolve(authorizationHeader: String?): Long {
        if (authorizationHeader == null || !authorizationHeader.startsWith("Bearer ")) {
            throw TaskError.unauthorized()
        }
        val token = authorizationHeader.substring("Bearer ".length)

        val claims = try {
            dispatcher.verify(token)
        } catch (e: VerifyException) {
            throw TaskError.unauthorized()
        }

        val userId = if (Dispatcher.isLocalIssuer(claims.iss)) {
            val id = claims.sub.toLongOrNull() ?: throw TaskError.userNotProvisioned()
            val user = repository.findUserById(id) ?: throw TaskError.userNotProvisioned()
            user.id
        } else {
            val user = repository.findUserByKeycloakSub(claims.sub) ?: throw TaskError.userNotProvisioned()
            user.id
        }
        // LOG_LEVEL=debugのときのみ出るリクエスト単位の詳細ログ(要約行はMain.ktのINFOログが担う)
        log.debug("resolved user_id={} from iss={} sub={}", userId, claims.iss, claims.sub)
        return userId
    }
}
